Fix FibreChannel volume plugin corrupting filesystem on detach

jsafrane opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
FibreChannel volume plugin misses one important step when removing a device: `multipath -f`, that flushes all multipath buffers to its individual paths. Without it, a filesystem on the device may get corrupted.

Calling `multipath -f` is easy, however, the patch grew a bit by passing all the necessary interfaces down to fc_util.go. The structure copies iscsi_util.go

See https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/storage_administration_guide/removing_devices for details.


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #97014

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed FibreChannel volume plugin corrupting filesystems on detach of multipath volumes.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jsafrane committed:
Fix FibreChannel volume plugin corrupting filesystem on detach

FibreChannel volume plugin misses one important step when removing a
device: "multipath -f". It flushes all multipath buffers to its individual
paths. Without it, a filesystem on the device may get corrupted.

jsafrane commented:
/priority important-soon

jingxu97 commented on pkg/volume/fc/fc_util.go:
```diff
@@ -373,6 +380,15 @@ func (util *fcUtil) DetachBlockFCDisk(c fcDiskUnmapper, mapPath, devicePath stri
+func (util *fcUtil) deleteMultipathDevice(exec utilexec.Interface, dmDevice string) error {
+	out, err := exec.Command("multipath", "-f", dmDevice).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to flush multipath device %s: %s\n%s", dmDevice, err, string(out))
```
just want to confirm whether in certain cases, it should continue instead of return error (e.g., the error means device no longer exist?)

jsafrane replied:
Good catch! That's why iscsi volume plugin ignores the error code...

jsafrane replied:
If the device is missing:

* Filesystem volumes: operation_generator.go will stop with "The path isn't device path or doesn't exist. Skip checking device path: /dev/dm-0" and it won't call volume plugin at all. Therefore, if the multipath device is removed, but the individual paths are still there, nothing will remove them.

* Block volumes: if the device is missing, the next `DetachBlockFCDisk` will stop in `checkPathExists`, if not earlier in `operation_generator.go` (did not test it).
https://github.com/kubernetes/kubernetes/blob/6f8bdb7cc1e28585edb5d47212ffb4d0ab3bf546/pkg/volume/fc/fc_util.go#L301

iscsi volume plugin actually ignores the error code from "multipath -f", continuing deleting the device paths. Which may result in volume corruption, if the multipath device still exists.. Now is the question what is actually correct - try to remove the individual paths to fully detach the devices and risk corruption OR do not remove the paths at all? The paths should just sit in /dev, not used by anything.

My gut feeling is to copy iscsi approach and ignore the exit code.

rootfs replied:
according to manpage
`-f     Flush (remove) a multipath device map specified as parameter, if unused.`

So if `multipath -f` fails, there is a good chance the device is still being used and removing it could cause data corruption. In this sense, the implementation in iscsi is indeed risky.

rootfs commented:
/approve

rootfs commented on pkg/volume/fc/fc_util.go:
```diff
@@ -373,6 +380,15 @@ func (util *fcUtil) DetachBlockFCDisk(c fcDiskUnmapper, mapPath, devicePath stri
+func (util *fcUtil) deleteMultipathDevice(exec utilexec.Interface, dmDevice string) error {
+	out, err := exec.Command("multipath", "-f", dmDevice).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to flush multipath device %s: %s\n%s", dmDevice, err, string(out))
```
nit: 
```go
return fmt.Errorf("failed to flush multipath device %s: %v\n%s", dmDevice, err, string(out))
```

jsafrane replied:
No, `%s` is fine for `errror` type.

gnufied commented on pkg/volume/fc/fc_util.go:
```diff
@@ -354,6 +358,9 @@ func (util *fcUtil) DetachBlockFCDisk(c fcDiskUnmapper, mapPath, devicePath stri
 	if len(dm) != 0 {
```
Same link - https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/storage_administration_guide/removing_devices#:~:text=Run%20multipath%20%2Dl%20command%20to,to%20remove%20the%20multipath%20device. recommends that for raw block devices we should call ` blockdev --flushbufs device`  before removing the device. Should we do that as well?

jsafrane replied:
Added as a separate commit.
BTW, we should refactor FC and iSCSI to use the same flush / device deletion. But that's for a separate PR.

gnufied replied:
yeah lets file that as a separate issue I think.

tsmetana commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

jsafrane committed:
Flush FibreChannel devices before deleting

If a FibreChannel device is used as a block volume, we should flush its I/O
before deleting its device. It is not strictly necessary when it's used as
a filesystem (mount), but it won't hurt either.

gnufied commented:
/retest
/lgtm
